api/src/main/java/jakarta/servlet/WriteListener.java

```diff
@@ -40,7 +40,21 @@ public interface WriteListener extends EventListener {
     void onWritePossible() throws IOException;
 
     /**
-     * Invoked when an error occurs writing data using the non-blocking APIs.
+     * Invoked when an error occurs writing data after {@link ServletOutputStream#setWriteListener(WriteListener)} has been
+     * called. This method will be invoked if there is a problem while data is being written to the stream and either:
+     * <ul>
+     * <li>{@link ServletOutputStream#isReady()} has been invoked and returned false</li>
```

Comment:
I also really don't like the idea of doubling up error handling. The original proposal meant that you only had to implement `onError`, with this change you now need to handle errors thrown from the stream. Should we add a section that if `onWritePossible` throws `IOException` then the `onError` method is called? It seems like obvious behavior but I am not sure if it is called out anywhere. This would mean that if `write` throws you can just let the exception propagate and the listener will handle it.

Reply:
Stuart, by `onError` in your comment, I'm assuming that you mean `AsyncListener.onError` and not `WriteListener.onError`.

So I do like the "if `OWP` throw then the `AL.onError` method is called" as a good way to give control to the application about how write errors are reported to `AL.onError`.

I agree it there is something strange about not reporting a known error to `WL.onError` until `isReady()` is called.  But if we do not, then we do not have an easy way to tell if a previous write has completed or not.  Currently the only way we have on knowing a previous write has completed is if `isReady()` has returned true.   Perhaps you could also say that if `onWritePossible` has been called, that also indicates completion... but really `isReady()` should be called from within OWP and checked for a true response to protect from spurious wakeups.   If `WL.onError` can be called at any time, then it may be called simultaneously to another thread calling `isReady()` and then the app will never know if the call to `WL.onError` was the result of the false return from `isReady` or if it just happened anyway and another call is on its way.

I'm not sure there is a good solution given the current API.  I think the best we can do is be rigorous with the scheduling so we at least avoid races like the one above.

Note that if the app you described really wants  to know about an error before the next  write is ready, there is nothing stopping it calling `isReady()` immediately after the `write` and then it will know that the operation has either completed or that either ODA or WL.onError will be called as soon as the operation is complete.

Reply:
Thinking about it some more I agree with your concerns about thread safety. The only way I can think to make this work is to allow this use case via the flush method. We could add something like 'If flush() is called in async mode then isReady must not return true until the data is written out to the client'. 

Then if you really care about error notification and are not going to immediately write again you can call `flush + isReady` to see the results.